### ISSUE TITLE
Proposal of a full-width userbar instead of "edit bird"

### ISSUE THREAD

**Issue by alexgleason** [torchbox/wagtail#2164]
I feel like there's a better way to do this. WordPress for instance puts a solid black bar at the top of your site when you're logged in, pushing the whole rest of the site about 30px down. This way you don't have to worry about the userbar obstructing your mouse clicks. It's annoying to have to figure out the right place to put it on a site (usually not at the top) and even then it looks weird and isn't easily made to fit into the bottom corners.

I think a horizontal bar option (similar to WordPress) would be better. What do you all think?

**Comment by JoshBarr** [comment]
It's a good idea. Pretty unobtrusive if it's only 30px. 

However, if your site includes a fixed position header, the user bar would overlap it. Any idea how Wordpress deals with that?

**Comment by alexgleason** [comment]
@JoshBarr Like this:

```css
html {
    margin-top: 32px !important;
}

#wpadminbar {
    height: 32px;
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    min-width: 600px;
    z-index: 99999;
}
```

Essentially just pushes the whole site down by 32px to create a gap, then puts the bar in the bap.

**Comment by JoshBarr** [comment]
It won't push down any `position: fixed` elements, but at least the massive z-index should help the admin bar stay on the top :+1:

**Comment by alexgleason** [comment]
Ah yeah, good point. I've never had a site collide with it before. But in the rare case it can be worked around with something like

```css
html.admin-mode .my-fixed-obj {
    top: 32px;
}
```

Or if you were really paranoid use some JS that checks for the existence of the bar then loops through every element giving it a margin-top of + 32px if it's position is fixed. Not recommending, just sayin'.

**Comment by davecranwell** [comment]
I think a bar the full width of the site, even if it's only 30px high, seems a bit overkill for just 3 buttons (5 during moderation), but I'd be open to ideas. Having the bar where it currently is _does_ sometimes get in the way of a search box etc, but moving it is trivial.

I too had imagined something like wordpress or drupal's bar when Wagtail was young, but in the event, there was so little to add to it, there was no need for it. I'm also mindful of how intrusive even a small bar can be on the brand of the resulting site. I'd also be interested to avoid the bar approach purely from the over-familiarity aspect - it's what everyone does.

For a while I've been considering giving the userbar more intelligence, so if it were placed in the bottom right, the buttons would automatically expand upwards, not down. Or if placed on the left, the buttons would appear to the right etc. A more delightful approach, basically, is what I'm after, rather than Yet Another Menu Bar.

**Comment by JoshBarr** [comment]
I think the iFrame might be our culprit. Right now, it has to be large enough to contain all the buttons, which means that it has quite a large hit area that can get in the way of the page beneath it:

![screen shot 2016-01-30 at 3 23 39 pm](https://cloud.githubusercontent.com/assets/1715456/12693047/88ea9cb8-c765-11e5-924c-ad7838002ccd.png)

If there were a way of reducing the size of the iframe, or dynamically resizing it via postmessage or something, that could help!

**Comment by alexgleason** [comment]
Why does it use an iframe anyway? I'm sure we could just make it not exist in an iframe to reduce that giant gap. That is definitely part of the problem.

**Comment by JoshBarr** [comment]
I imagine it's to ensure the styling and assets for the userbar aren't clobbered by other things happening on the page (eg, !important styles, JS that adds classes to all <divs>, etc). Could also be to do with the cross-domain policy for multi-site implementations? I'm not 100% sure on the reasons.

It could be re-built without the need for an iframe, I think. Use JS to inline all the styles + serve the icons as base64 encoded SVGs inline. The target region could also be reduced to just the initial square, with a tap to show the editing options (more tablet friendly). 

Only, we don't want to break any customisations people do to the widget currently. Do we know how much the admin bar gets customised by implementers?

**Comment by m1kola** [comment]
Did you saw django-debug-toolbar? It provides toolbar similar to Wagtail's userbar and gives ability to change position on right side of window. Django-debug-toolbar writes position (from top right corner) into cookie.

<img width="398" alt="screen shot 2016-01-30 at 16 08 35" src="https://cloud.githubusercontent.com/assets/509198/12695869/dc42ab0c-c76b-11e5-90c8-9405029c2d2f.png">

**Comment by davecranwell** [comment]
Yes the iframe was to avoid undesirable effects of the implementation's own stylesheet, but if there were a way to ensure everything works without needing an iframe, that would solve all our problems.

**Comment by JoshBarr** [comment]
Sweet! I think this can be done with some very careful styles (a mix of inline and `!important` probably). I'll see if any of my team have time to look into it.

**Comment by JoshBarr** [comment]
Update on this one: @mr-winter (Thomas) is taking a look at the userbar styles. 

The approach taken is something like the attached image. We looked through a bunch of sites on Made with Wagtail, and concluded that the majority of them use some form of top navigation bar. Using the bottom corner of the screen helps the Wagtail UI stay out of the way of the primary site nav, while not overlapping critical content. 

![pasted image at 2016_02_02 08_53 am](https://cloud.githubusercontent.com/assets/1715456/12777832/2aa4804e-cac4-11e5-82c8-2f30749db57d.png)

**Comment by davecranwell** [comment]
Nice! like it!

**Comment by ollywillans** [comment]
+1 :-)

On 3 February 2016 at 09:48, Dave Cranwell <notifications@github.com> wrote:

> Nice! like it!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/torchbox/wagtail/issues/2164#issuecomment-179131256>.
>



-- 

Olly Willans
Director

Torchbox Ltd | Southill Park | Charlbury | OX7 3EW

Direct: +44 (0)1608 811 921
Office: +44 (0)1608 811 870
Email: olly.willans@torchbox.com
http://torchbox.com

**Comment by timorieber** [comment]
+1

**Comment by benebun** [comment]
Related: We had to disable the userbar because we don't use https in the frontend:
- we redirect /admin to https
- when editors visit the frontend via http this redirect interferes with the loading of the userbar so it displays an empty iframe that invisibly overlaps the UI of the website.
Should I file a separate issue for that? An alternative workaround would probably be to use https everywhere…

**Comment by gasman** [comment]
@benebun Yes, it sounds like it's worth opening a new issue about that. Thanks!

**Comment by alexgleason** [comment]
I'd like to revisit the idea of a top menubar with an example.

![screenshot from 2016-02-04 11 16 39](https://cloud.githubusercontent.com/assets/3639540/12822308/7e1f2ad2-cb35-11e5-8927-c9c6746e366e.png)

(Note that "My Site" in the menubar would be whatever the name of the website is)

Here is the menubar in WordPress. A little more cluttered than it needs to be, sure.
 
![screenshot from 2016-02-04 11 19 40](https://cloud.githubusercontent.com/assets/3639540/12821565/e7445ae0-cb31-11e5-9cf5-42dcc75cd69d.png)

But one nice thing is how each of these items can have dropdown options. This would allow for an intuitive way to implement #2195.

![screenshot from 2016-02-04 11 18 42](https://cloud.githubusercontent.com/assets/3639540/12821615/1599ed9c-cb32-11e5-8828-3b7ebd33f409.png)

Also, you can use it to get to the dashboard. The current implementation can also do that. But watch what happens once you're inside the dashboad:

![screenshot from 2016-02-04 11 34 03](https://cloud.githubusercontent.com/assets/3639540/12821851/3beebd82-cb33-11e5-8e25-325c24344871.png)

This is where I think the difference is important. The menubar stays at the top of the page, even inside of the dashboard, offering a consistent user experience throughout managing the site. Whether on the front-end, or in the admin panel, you can access the menubar for common managerial tasks. By the way, you can get back to the site from the admin in the same way you got the the admin from the site! Thereby implementing #2047 

![screenshot from 2016-02-04 11 20 37](https://cloud.githubusercontent.com/assets/3639540/12821724/97f90db8-cb32-11e5-8b1b-62906c3cafa3.png)

Here are some of the benefits:

* Wide horizontal space:
    * We can easily add new base options to the menubar in the future
    * Sites could be configured on a per-site basis to add new items to the menubar. Some sites may need more configuration to be at the fingertips of the content manager.
    * Plugins could add items to the menubar through a Wagtail hook.
* Always present, but contextual.
    * It's intuitive to let the user always access controls to get in and out of the admin !important
    * Other items can change depending on where exactly you are. One set of controls for pages, a different set for the dashboard.
    * This could even give us the ability to configure custom menu items when inside certain page types on the front-end. It's a pipe dream, but at least a menubar makes it possible.

**Comment by davecranwell** [comment]
I can't refute any of your suggestions for why this style of bar would work, they're all valid. However the fact remains that this isn't a small change and it would require a great deal more thought (than the design suggested by @mr-winter/@joshbarr). There are benefits to a bar like yours, but it also introduces more moving parts, more problems for integration, a greater dependency on serious UX and design work, and a need for alterations to most of Wagtail itself.

Our mantra with Wagtail so far has been to alter something only when it's _needed_. The difficulties with the current userbar placement and iframe-shenanigans seem to merit tweaks, but I wouldn't say there's been quite enough clamour for the change on the scale you're suggesting.

**Comment by alexgleason** [comment]
Well, I'm not suggesting that we implement any of those features, just that we implement our current features as a menubar. My above post was meant to express the flexibility that a menubar can offer, simply due to the size and shape of it. I believe our philosophies are in line. But why not lay the groundwork in a way that's more flexible and conducive to additions while we're fixing a problem that currently exists?

**Pros:**
* Fixes the problem
* _Always_ stays out of the way
* Offers greater theoretical potential

**Cons:**
* Menubar will have empty space

This is the way I view it; the empty space is worth it to me. I do like the circular design presented above, but being in the bottom corner doesn't guarantee it will stay out of the way—a true menubar actually pushes the whole site down to make room for itself absolutely guaranteeing it will not be in the way. As far as the empty space, I don't think it will make an impact. The bar can be solid black and white, and only about 30-35px tall so it blends in with any site.

**Comment by heymonkeyriot** [comment]
I think the idea of the top-bar has a lot going for it. As Alex says it has a big advantage of consistency. Contextually, I think being a top bar allows it to sit more easily within the architecture of the browser, which will lead to less interference with the look and feel of the actual website being viewed.

The suggestion so far with the top bar is to have dropdown menus, but I wonder if at desktop he space could also be used to ensure that the links, currently only revealed by the edit bird on hover, are available at all times. That would reduce friction of the user taking an action around edit/ publish etc. When the viewport becomes narrower they could placed within a menu?

**Comment by JoshBarr** [comment]
Right now, I think the team have a good improvement to the existing adminbar that doesn't require a heap of design work, and won't result in anyone with a `position:fixed` header on their site having to re-work it when the wagtail admin bar is present. 

Might i suggest that we make a top menu strip a switchable config option, and look at designing that properly after we've got this next iteration of the floating admin button merged in?

**Comment by davecranwell** [comment]
Yeah, I think my biggest concern with any UI change, is waiting so long for the "ultimate" solution that we never move at all.

**Comment by JoshBarr** [comment]
Also, if you look at where Wordpress are heading, they're decoupling the admin entirely from the front end. Calypso is a desktop app (!), so no admin toolbar.

**Comment by alexgleason** [comment]
^ I agree with both of you. :+1:

**Comment by davecranwell** [comment]
We should bear in mind that Wagtail has a relatively tiny user-base currently. We're certainly no Wordpress or Drupal. While I don't want to advocate rash decisions, we _are_ still young enough to make some mistakes provided we iterate past them onto something better....assuming we can iterate rapidly enough.

**Comment by MechanisM** [comment]
Personally, I don't like how it is done in WordPress. I would prefer this one:
![image](https://cloud.githubusercontent.com/assets/121781/12830296/9122714c-cb9f-11e5-8247-88f98003a8a0.png)
or current one(but without iframe).

**Comment by jonnyscholes** [comment]
I'd be very keen to see the new floating version of the userbar. That said, we have a site we've worked on recently that uses the same material design inspired menu buttons where the bottom-right location would clash. Would it be worth having an option available to developers to choose which corner it sits in? It wouldn't dramatically increase the dev time and would mean a developer can ensure it's never in a spot that clashes with the design.

**Comment by JoshBarr** [comment]
Yeah positionable in any corner would be good. It's not very much work either. I think we'll add that feature and an associated setting.

**Comment by gasman** [comment]
Implemented in #2201
